store/postgres/asset_repository.go

```diff
@@ -71,59 +92,213 @@ func (r *AssetRepository) GetCount(ctx context.Context, config asset.Config) (to
+			a.service as"service",
+			a.description as "description",
+			a.data as "data",
+			a.labels as "labels",
+			a.version as "version",
+			a.created_at as "created_at",
+			a.updated_at as "updated_at",
+			a.changelog as "changelog",
+			a.owners as "owners",
+			u.id as "updated_by.id",
+			u.email as "updated_by.email",
+			u.provider as "updated_by.provider",
+			u.created_at as "updated_by.created_at",
+			u.updated_at as "updated_by.updated_at"
+		FROM assets_versions a
+		LEFT JOIN users u ON a.updated_by = u.id
+		WHERE
+			a.asset_id = $1
+		ORDER BY version DESC
+		LIMIT $2
+		OFFSET $3
+	`, id, size, cfg.Offset)
+	if err != nil {
+		err = fmt.Errorf("failed fetching last versions: %w", err)
+		return
+	}
+
+	if len(assetModels) == 0 {
+		err = asset.NotFoundError{AssetID: id}
+		return
+	}
+
+	for _, am := range assetModels {
+		av, ferr := am.toAssetVersion()
+		if ferr != nil {
+			err = fmt.Errorf("failed converting asset model to asset version: %w", ferr)
+			return
+		}
+		avs = append(avs, av)
+	}
+
+	return avs, nil
+}
+
+// GetByVersion retrieves the specific asset version
+func (r *AssetRepository) GetByVersion(ctx context.Context, id string, version string) (ast asset.Asset, err error) {
+	if !isValidUUID(id) {
+		err = asset.InvalidError{AssetID: id}
+		return
+	}
+
+	latestAsset, err := r.GetByID(ctx, id)
+	if errors.Is(err, sql.ErrNoRows) {
+		err = asset.NotFoundError{AssetID: id}
+		return
+	}
+
+	if err != nil {
+		return
+	}
+
+	if latestAsset.Version == version {
+		ast = latestAsset
+		return
+	}
+
+	var assetModel AssetModel
+	err = r.client.db.GetContext(ctx, &assetModel, `
+		SELECT
+			a.asset_id as "id",
+			a.urn as "urn",
+			a.type as "type",
+			a.name as "name",
+			a.service as"service",
+			a.description as "description",
+			a.data as "data",
+			a.labels as "labels",
+			a.version as "version",
+			a.created_at as "created_at",
+			a.updated_at as "updated_at",
+			a.changelog as "changelog",
+			a.owners as "owners",
+			u.id as "updated_by.id",
+			u.email as "updated_by.email",
+			u.provider as "updated_by.provider",
+			u.created_at as "updated_by.created_at",
+			u.updated_at as "updated_by.updated_at"
+		FROM assets_versions a
+		LEFT JOIN users u ON a.updated_by = u.id
+		WHERE a.asset_id = $1 AND a.version = $2`, id, version)
+
+	if errors.Is(err, sql.ErrNoRows) {
+		err = asset.NotFoundError{AssetID: id}
+		return
+	}
+
+	if err != nil {
+		err = fmt.Errorf("failed fetching asset version: %w", err)
+		return
+	}
+
+	ast, err = assetModel.toVersionedAsset(latestAsset)
 
 	return
 }
 
 // Upsert creates a new asset if it does not exist yet.
 // It updates if asset does exist.
 // Checking existance is done using "urn", "type", and "service" fields.
-func (r *AssetRepository) Upsert(ctx context.Context, ast *asset.Asset) (string, error) {
-	assetID, err := r.getID(ctx, ast)
+func (r *AssetRepository) Upsert(ctx context.Context, userID string, ast *asset.Asset) (string, error) {
```

Comment:
Since the updated by is coming from another source (not from ast object), distinguishing it should give more clearer function signature. Although the proper way to have that signature is in service layer. But we don't have service layer for this domain.

But I think I agree with you, the repository could just have arguments like that.

Reply:
I see, if we have a service, we can also add a validation like "`ast.UpdatedBy.ID` is required" there. But yeah, leaving the decision to you for this